### ISSUE TITLE
SQL dependencies >= 100 seconds are not tracked

### ISSUE THREAD

**Issue by taddison** [Microsoft/ApplicationInsights-dotnet-server#1133]
We noticed a few long-running SQL calls were not being tracked by app insights, and after investigating found the cutoff to be 100 seconds.

The below app will log dependency calls for 98,99 seconds and nothing for 100,101.  We repeated this several times to try and isolate any timing issues and got very consistent results.

Setup:
- netcoreapp2.2
- Microsoft.ApplicationInsights 2.8.1
- MIcrosoft.ApplicationInsights.DependencyCollector 2.8.1

(on GitHub [here](https://github.com/taddison/csharp-coffre/tree/7ece318face449639b36ed82ee3ffa180c5a4fc0/appinsights-sql-test), needs an instrumentation key and SQL database to work).

```csharp
using System;
using Microsoft.ApplicationInsights;
using Microsoft.ApplicationInsights.Extensibility;
using System.Threading;
using System.Data.SqlClient;
using Microsoft.ApplicationInsights.DependencyCollector;

namespace appinsights_sql_test
{
    class Program
    {
        static void Main(string[] args)
        {
            var config = TelemetryConfiguration.Active;
            config.InstrumentationKey = "###";
            var module = new DependencyTrackingTelemetryModule();
            module.Initialize(config);

            var telemetryClient = new TelemetryClient();
            telemetryClient.TrackTrace("Start Benchmark");

            using(var conn = new SqlConnection("server=localhost;initial catalog=master;integrated security=SSPI"))
            {
                conn.Open();
                for(var i = 98; i <= 102; i++)
                {
                    var ts = new TimeSpan(0, 0, i);
                    using (var cmd = new SqlCommand($"waitfor delay '00:{ts.Minutes:00}:{ts.Seconds:00}'",conn))
                    {
                        client.TrackTrace($"Running SQL query for {ts.TotalSeconds} seconds");
                        cmd.ExecuteNonQuery();
                    }
                }
            }

            Console.WriteLine("Done.");
            telemetryClient.Flush();
            // https://github.com/Microsoft/ApplicationInsights-dotnet/issues/407
            Thread.Sleep(10000);
        }
    }
}
```

Here is an example of the captured telemetry from earlier tests - you can see we no longer get dependency data from 100 seconds.

```kql
union traces, dependencies
| where timestamp > todatetime("20181220 21:02:00") and timestamp < todatetime("20181220 21:16:00")
| project timestamp, message, duration
| order by timestamp desc
```

![image](https://user-images.githubusercontent.com/2743558/50353499-7614ba00-0516-11e9-80e3-614507c4c810.png)

**Comment by cijothomas** [comment]
@taddison Will exposing the max duration to be set by user solve your issue?

**Comment by taddison** [comment]
Yes - though we'd curious to know what kind of impact raising this would have to make that call/tradeoff.

For context, most of our calls are extremely fast, and the places where we have gaps are typically the intersection of:

- Application startup (long-running query to do a one-time population of caches that are incremental once warmed)
- Dev/Test environment (hardware isn't as good/someone is doing terrible things to the SQL instance)

This first came to our attention when we did a pass to try and better understand why some Dev/Test scenarios were so slow, and the absence of the SQL dependencies we knew should have been stood right out.

**Comment by cijothomas** [comment]
DependencyCollector keeps an internal cache to correlate between the 'begin' and 'end' of a SQL call. Items are entered to cache upon Begin, and are looked up and removed upon End of calls.
Also this cache auto clears entries after timeout of 100 secs now - this is done to make sure that cache does not grow indefinitely when End never arrives. (due to bugs !).

Increasing the cache expiration to higher values means that in the event of End not receiving, cache will be kept longer, possibly causing memory pressure.

**Comment by taddison** [comment]
We'd definitely appreciate being able to test - either by environment (e.g. set it high in dev/test), or even by server (e.g. server1 high vs. server2 low in prod) to gauge impact.  If ultimately the impact is low and the cache doesn't bloat, we'd always rather capture the long-running queries than have them go missing.  The ones that take a long time are often the ones we really want to focus on!
